### ISSUE TITLE
[BUG] Latest Pi-hole update - Interface listening behavior

### ISSUE THREAD

**Issue by pchristod** [pivpn/pivpn#1192]
## In raising this issue, I confirm the following: 

`{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pivpn/pivpn/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting can be *is* directly related to the pivpn installer script.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pivpn/pivpn/wiki/FAQ), [closed issues](https://github.com/pivpn/pivpn/issues?q=is%3Aissue+sort%3Aupdated-desc+is%3Aclosed), and [open issues](https://github.com/pivpn/pivpn/issues?q=is%3Aissue+sort%3Aupdated-desc+is%3Aopen)).

**Describe the bug**
Hi,

with the latest Pi-hole Update from like 2 days ago (https://pi-hole.net/2020/11/28/pi-hole-core-web-v5-2-and-ftl-v5-3-released/#page-content), it seems not valid anymore to listen simply on all interfaces, like the PiVPN installer sets it automatically. Corresponding Commit https://github.com/pivpn/pivpn/commit/139f16594d41ab49991c1307ec7399ec60d09db7 to what I mean.
It seems that actually it is (now?) needed to "Listen on all Interfaces and permit all origins" like specified in a previous Commit (https://github.com/pivpn/pivpn/commit/5b2bc9ba70a24250246da8c942a31e123a431096) which was superseded by the above Commit. Naturally I'm a fan of having the Pi security wise as tight as possible. I know that the Option to listen everywhere + permit all origins should technically be secure while the Pi is within the LAN but I found the approach you set in your installer script better (be it for the peace of mind). Why expose more if it is not necessarily needed? 

I updated this afternoon to the latest Pi-hole release rendering me unable to resolve anything, connecting to the VPN Server itself however is working fine from the clients. I reinstalled PiVPN to have a clean working copy but only changing the listening behaviour did enable my clients to actually load/send data. Can anyone with the latest Pi-hole release and PiVPN check and possibly confirm that? 

**To Reproduce**
1. Update to latest Pi-hole release
2. Reboot
3. Clients should be able to connect but have no internet access (also nothing will be in the Pi-hole query logs itself)

**Expected behavior**
Clients should be able to have internet access when pihole is used

**VPN Server (please complete the following information):**
- Wireguard
- Raspbian Stretch
- latest Pi-hole Release
- latest PiVPN Release

**Comment by blackm** [comment]
I am having similar issues, setting to _Listen only on interface eth0_ is working for local (wlan) clients with vpn off and vpn clients.

**Comment by newAccount51** [comment]
FWIW, I haven't been able to replicate this. I recently installed the latest release of PiVPN wg on top of PiHole, after which my PiHole has updated to the latest 5.2/5.3 version. I checked the interface options and it's set to the top option: listen locally.

**Comment by Kopernikus1979** [comment]
Installed PiVPN yesteday and got same issue, then I saw this post seems during the PiVPN install the listen setting in PiHole gets changed to listen all interfaces... the way I fixed it and this is how I managed my PiHole who is MultiHomed is setting PiHole to listen to the eth0 interface only and then in the dnqmasq.d define all the interfaced that Pi-Hole needs listening to so in my case this is eth0, eth0.10, eth0.20, eth0.30 and to make PiVPN work also wg0

**Comment by pchristod** [comment]
Yes, if I'm not mistaken this was the previous solution as to how PiVPN had that implemented, the interface was added to /etc/dnsmasq.d/02-pivpn.conf via "interface=tun0" for example. However it was not recommended by PiHole devs to drop a static configuration file during PiVPN installation with an interface set into the users directory, as seen per discussion here https://discourse.pi-hole.net/t/connection-problem-after-4-3-3-4-3-5-update-with-pivpn/28708/25

Eventually it got changed in the PiVPN Installer as per the Commits above to the solution that there is right now (PiVPN Installer sets the Interface Listening behaviour to the top option, which is also the Pi-Hole default). 

I'm not saying that this is definately an issue, just opened this topic to confirm if someone ran into the same problem. I see that also two other comments are added here, would be interesting to know if you guys have default installation or changed anything. Most importantly in /etc/dnsmasq.d/02-pivpn.conf. Maybe there are even entries from old pivpn installation in there. My current output looks like this, which should be the stock created by the installer:

`addn-hosts=/etc/pivpn/hosts.wireguard`

I can only say that (at least) I'm not able to use PiVPN in combination with Pi-Hole if I leave the listening behaviour in Pi-Hole to the value the PiVPN installer sets.

**Comment by orazioedoardo** [comment]
Basically when PiVPN installer runs `pihole -a -i local` the wg0 interface is not up yet, so dnsmasq does not recognize queries from 10.6.0.0/24 as "local" and you get `Ignoring query from non-local network` in the pi-hole log. If you change the listening behavior after starting wireguard or run `pihole restartsdns` at any time wireguard is up, it sorts out. As a workaround, the installer can just defer the pi-hole tweak at the end of the installation.

**Comment by pchristod** [comment]
That makes a lot of sense. This race condition seems to be also happening when restarting the PI then if course.
I've got to try that later but basically in any case one needs to ensure the interface is started before the pihole dns service.

**Comment by orazioedoardo** [comment]
> This race condition seems to be also happening when restarting the PI then if course.

Yes definitely, in my test below it didn't, but that could happen depending on how many services your system is expected to start.

```
pi@raspberrypi:~ $ journalctl -b --no-pager | grep -E 'wireguard|pihole-FTL'
Dec 08 13:23:21 raspberrypi systemd[1]: Starting LSB: pihole-FTL daemon...
Dec 08 13:23:21 raspberrypi pihole-FTL[526]: Not running
Dec 08 13:23:21 raspberrypi wg-quick[527]: [#] ip link add wg0 type wireguard
Dec 08 13:23:21 raspberrypi kernel: wireguard: WireGuard 1.0.20201112 loaded. See www.wireguard.com for information.
Dec 08 13:23:21 raspberrypi kernel: wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
Dec 08 13:23:22 raspberrypi pihole-FTL[526]: FTL started!
Dec 08 13:23:22 raspberrypi systemd[1]: Started LSB: pihole-FTL daemon.
```

> Edit: I tried to verify what you were saying and it seems you are correct. When the DNS Service is restarted during runtime (when the interface is up) then it works with the default Pi-Hole option. However I could not get this to work during restart. I noticed Pi-Hole offers an option "DEFAULT_STARTUP=" which delays startup for such scenarios. While the option is working, not even 45 seconds are enough for the Wireguard Interface to come up, which seems kind of long. Haven't tried it further at the moment.

How about adding an override to the wg-quick unit to start before pihole-FTL? Run `sudo systemctl edit wg-quick@wg0.service`, past the following and reboot:
```
[Unit]
Before=pihole-FTL.service
```

**Comment by pchristod** [comment]
Very nice, you are a lifesaver @orazioedoardo !
Great solution, I didn't even know you could do that. This is working great and now everything is up when the DNS service is started.

```
pi@raspberrypi2:~ $ uptime -s
2020-12-08 13:49:52

pi@raspberrypi2:~ $ systemctl status wg-quick@wg0.service
● wg-quick@wg0.service - WireGuard via wg-quick(8) for wg0
   Loaded: loaded (/lib/systemd/system/wg-quick@.service; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/wg-quick@wg0.service.d
           └─override.conf
   Active: active (exited) since Tue 2020-12-08 13:50:57 CET; 1min 7s ago
     Docs: man:wg-quick(8)
           man:wg(8)
           https://www.wireguard.com/
           https://www.wireguard.com/quickstart/
           https://git.zx2c4.com/wireguard-tools/about/src/man/wg-quick.8
           https://git.zx2c4.com/wireguard-tools/about/src/man/wg.8
  Process: 1294 ExecStart=/usr/bin/wg-quick up wg0 (code=exited, status=0/SUCCESS)
 Main PID: 1294 (code=exited, status=0/SUCCESS)

pi@raspberrypi2:~ $ systemctl status pihole-FTL
● pihole-FTL.service - LSB: pihole-FTL daemon
   Loaded: loaded (/etc/init.d/pihole-FTL; generated; vendor preset: enabled)
   Active: active (exited) since Tue 2020-12-08 13:50:58 CET; 1min 12s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 1380 ExecStart=/etc/init.d/pihole-FTL start (code=exited, status=0/SUCCESS)
```

Before you wrote your post I also went down this route to try and find out when wg-quick is run after a reboot so I set pihole delay back to 0. It ran 1 minute and 4 seconds after the DNS service was started. So the seconds solution would probably be to add a delay for a little over 1 minute, which would be pretty bad...

Thanks for sharing your knowledge and that solution. Wonder why that came up after the Pihole update but it might just be a coincidence as I rarely reboot my pi (once or twice every year) so it might just be that I never noticed before.

**Comment by jojolll** [comment]
> ```
> [Unit]
> Before=pihole-FTL.service
> ```

Hello

I wanted to inform you that I also solved my problem with this technique. 
Yesterday I performed a clean install of PiHole and then PiVPN, on a fresh RaspberryPi OS on SSD (Pi 4).

Installing PiHole and configuring it is absolutely no problem at all. Everything works.
The PiVPN installation goes well too, and everything works once installed.

However, every time I reboot after these installations, things get more complicated: The raspberry pi can't access the internet anymore. and the only solution to solve the problem (until the next reboot) was to physically unplug the ethernet cable, then plug it back in and everything is back to "normal" (incomprehensible knowing that the pi has a fixed IP, this shouldn't change anything). Moreover, restarting the interface by software does not solve the problem.
I also noticed that at some reboots (about one time out of 5) no network problems occurred... 

I also redid an installation with the test branch, exactly the same behavior.


Once I came across this thread I tried the orazioedoardo technique and it seems to work every time I reboot.
Could it be that PiHole became too fast to start for WireGuard in its latest version?
